otlp/metrics.go

```diff
@@ -0,0 +1,372 @@
+	strs := []string{
+		resourceKey,
+		strconv.Itoa(int(ts.UnixNano())),
+	}
+
+	labels := dp.GetLabels()
+	for i := range labels {
+		strs = append(strs, fmt.Sprintf("%s=%s", labels[i].Key, labels[i].Value))
+	}
+
+	attributes := dp.GetAttributes()
+	for i := range attributes {
+		strs = append(strs, fmt.Sprintf("%s=%s", attributes[i].Key, attributes[i].Value.String()))
+	}
+
+	return strings.Join(strs, ",")
+}
+
+// repackageDeprecatedOTLPIntDataPoints converts an IntDataPoint into a NumberDataPoint. This allows us to use
+// the same code to handle both types later on.
+func repackageDeprecatedOTLPIntDataPoints(dataPoints []*metrics.IntDataPoint) (wrapped []*metrics.NumberDataPoint) {
+	for _, dp := range dataPoints {
+		wrapped = append(wrapped, &metrics.NumberDataPoint{
+			Labels:       dp.GetLabels(),
+			TimeUnixNano: dp.GetTimeUnixNano(),
+			Value:        &metrics.NumberDataPoint_AsInt{dp.GetValue()},
+			// Note: we're not saving StartTimeUnixNano or Exemplars, so drop this data here
+		})
+	}
+	return wrapped
+}
+
+// repackageDeprecatedOTLPIntHistogramDataPoints converts an IntHistogramDataPoint into a HistogramDataPoint.
+// This allows us to use the same code to handle both types later on.
+func repackageDeprecatedOTLPIntHistogramDataPoints(datapoints []*metrics.IntHistogramDataPoint) (wrapped []*metrics.HistogramDataPoint) {
+	for _, dp := range datapoints {
+		val := float64(dp.GetSum())
+		wrapped = append(wrapped, &metrics.HistogramDataPoint{
+			Labels:         dp.GetLabels(),
+			TimeUnixNano:   dp.GetTimeUnixNano(),
+			Count:          dp.GetCount(),
+			Sum:            &val,
+			BucketCounts:   dp.GetBucketCounts(),
+			ExplicitBounds: dp.GetExplicitBounds(),
+			// Note: we're not saving StartTimeUnixNano or Exemplars, so drop this data here
+		})
+	}
+	return wrapped
+}
+
+var percentiles = map[string]float64{
+	".p001": 0.001,
+	".p01":  0.01,
+	".p05":  0.5,
+	".p10":  0.10,
+	".p25":  0.25,
+	".p50":  0.50,
+	".p75":  0.75,
+	".p90":  0.90,
+	".p95":  0.95,
+	".p99":  0.99,
+	".p999": 0.999,
+}
+
+func calculateQuantiles(count uint64, bucketCounts []uint64, bounds []float64) map[string]float64 {
+	quantiles := make(map[string]float64)
+	if len(bounds) > 0 && len(bucketCounts) > 0 {
+		var cumulativeTotal uint64
+		bucketQuantiles := make([]float64, 0, len(bucketCounts))
+		for _, bc := range bucketCounts {
+			cumulativeTotal += bc
+			bucketQuantiles = append(bucketQuantiles, float64(cumulativeTotal)/float64(count))
+		}
+
+		for key, percentile := range percentiles {
+			highestBound := bounds[0]
+			for i, quantile := range bucketQuantiles {
+				if quantile > percentile {
+					break
+				}
+
+				switch {
+				case i < len(bounds):
+					highestBound = bounds[i]
+				default:
+					// this is an odd scenario that we are not entirely sure creates
+					// valid data. We can use this to flag when investigation is needed
+					// potentially asking the customer for permission to look at the data
```

Comment:
Feels like you're missing a telemetry hook here?

Reply:
Yes, we are. I was thinking we would keep telemetry out of scope for this particular PR because we will likely need a way to emit a bunch of telemetry data for metrics. We don't seem to emit telemetry for traces & logs from Husky itself so we'll need a way to get telemetry data we care about for metrics out of Husky for shepherd to use.

Reply:
I don't mind if it's not part of this PR, but I don't think it's acceptable to push a change to Shepherd that will cause us to lose significant amounts of telemetry.

Maybe we can talk in the morning about finding a way to allow shepherd to inject a tracer into Husky.